src/pynwb/validation.py

```diff
@@ -0,0 +1,197 @@
+      - BuildManager object for opening the file for validation
+      - Dict with the full result from NWBHDF5IO.load_namespaces
+    """
+
+    catalog = NamespaceCatalog(
+        group_spec_cls=NWBGroupSpec, dataset_spec_cls=NWBDatasetSpec, spec_namespace_cls=NWBNamespace
+    )
+
+    if io is not None:
+        # TODO update HDF5IO to have .file property to make consistent with ZarrIO
+        # then update input arguments here
+        namespace_dependencies = io.load_namespaces(namespace_catalog=catalog, 
+                                                    file=io._file)
+    else:
+        from pynwb import _get_backend
+        backend_io = _get_backend(path, method=driver)
+        namespace_dependencies = backend_io.load_namespaces(namespace_catalog=catalog, 
+                                                            path=path, 
+                                                            driver=driver, 
+                                                            aws_region=aws_region)
+
+    # Determine which namespaces are the most specific (i.e. extensions) and validate against those
+    candidate_namespaces = set(namespace_dependencies.keys())
+    for namespace_dependency in namespace_dependencies:
+        candidate_namespaces -= namespace_dependencies[namespace_dependency].keys()
+
+    # TODO: remove this workaround for issue https://github.com/NeurodataWithoutBorders/pynwb/issues/1357
+    candidate_namespaces.discard("hdmf-experimental")  # remove validation of hdmf-experimental for now
+    cached_namespaces = sorted(candidate_namespaces)
+
+    if len(cached_namespaces) > 0:
+        type_map = TypeMap(namespaces=catalog)
+        manager = BuildManager(type_map=type_map)
+    else:
+        manager = None
+
+    return cached_namespaces, manager, namespace_dependencies
+
+@docval(
+    {
+        "name": "io",
+        "type": HDMFIO,
+        "doc": "An open IO to an NWB file.",
+        "default": None,
+    },  # For back-compatability
+    {
+        "name": "namespace",
+        "type": str,
+        "doc": "A specific namespace to validate against.",
+        "default": None,
+    },  # Argument order is for back-compatability
+    {
+        "name": "path",
+        "type": (str, Path),
+        "doc": "NWB file path.",
+        "default": None,
+    },
+    {
+        "name": "use_cached_namespaces",
+        "type": bool,
+        "doc": "Whether to use namespaces cached within the file for validation.",
+        "default": True,
+    },
+    {
+        "name": "verbose",
+        "type": bool,
+        "doc": "Whether or not to print messages to stdout.",
+        "default": False,
+    },
+    {
+        "name": "driver",
+        "type": str,
+        "doc": "Driver for h5py to use when opening the HDF5 file.",
+        "default": None,
+    }, 
+    returns="Validation errors in the file.",
+    rtype=(list, (list, bool)),
+    is_method=False,
```

Comment:
```suggestion
    is_method=False,
    allow_positional=AllowPositional.ERROR,
```
And
```
from hdmf.utils import AllowPositional
```
above.

I think we should require keyword arguments. What do you think? It's hard to make this change later.

Reply:
We could also set it to warn first:
```
    allow_positional=AllowPositional.WARNING,
```

Reply:
Agreed, I think that's a good idea. We can warn first to be consistent with all the other functions we're updating to warn in this release.